libs/src/api/Users.ts

```diff
@@ -993,36 +993,44 @@ export class Users extends Base {
     }
   }
 
+  /**
+   * Waits for the input replica set to be indexed by the discovery node
+   * If then replica set matches at the requested block number -> return null
+   * If the replica set response is null at the block number -> throw error
+   * If the replica set is mismatched at the block number -> throw error
+   * If the timeout is exceeded before replica set indexed -> throw error
+   */
   async waitForReplicaSetDiscoveryIndexing(
     userId: number,
     replicaSetSPIDs: number[],
     blockNumber: number,
     timeoutMs = 60000
-  ) {
+  ): Promise<void> {
     const asyncFn = async () => {
       while (true) {
         const encodedUserId = Utils.encodeHashId(userId)
+        let replicaSet
         try {
-          const replicaSet = await this.discoveryProvider.getUserReplicaSet({
+           replicaSet = await this.discoveryProvider.getUserReplicaSet({
             encodedUserId: encodedUserId!,
             blockNumber
           })
-          if (replicaSet) {
-            if (
-              replicaSet.primarySpID === replicaSetSPIDs[0] &&
-              replicaSet.secondary1SpID === replicaSetSPIDs[1] &&
-              replicaSet.secondary2SpID === replicaSetSPIDs[2]
-            ) {
-              break
-            } else {
-              throw new Error(
-                `[User:waitForReplicaSetDiscoveryIndexing()] Indexed block ${blockNumber}, but did not find matching sp ids`
-              )
-            }
-          }
         } catch (err) {
           // Do nothing on error
         }
+        if (replicaSet) {
+          if (
+            replicaSet.primarySpID === replicaSetSPIDs[0] &&
+            replicaSet.secondary1SpID === replicaSetSPIDs[1] &&
+            replicaSet.secondary2SpID === replicaSetSPIDs[2]
+          ) {
+            break
+          } else {
+            throw new Error(
+              `[User:waitForReplicaSetDiscoveryIndexing()] Indexed block ${blockNumber}, but did not find matching sp ids`
+            )
+          }
+        }
         await Utils.wait(500)
```

Comment:
i'm confused - this seems like it still will retry until timeout if replicaset is falsey?

Reply:
Well this is in a `await Utils.racePromiseWithTimeout(` so it will retry until that point in time
If the call `this.discoveryProvider.getUserReplicaSet` errors then blocknumber is not reached and it should wait and retry. 
If it returns with an object then the blocknumber was indexed - in that case if the replica matches then break out successfully, else throw an error to break out of the loop, and exit early

Reply:
k discussed offline - confusion here is that if no replica set found, the discovery call will return empty object, which is truthy

solution - improve docs